# *Reading the Data*

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Gene Expression File - Input variables

In [2]:
df_main = pd.read_csv("miller2005_genomicMatrix", sep='\t', header=None)

C:\Users\kvman\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3077: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251

In [3]:
#input(x values)
df_main.head()

0                    1                   2                   3    \
0        probe              X100B08             X101B88             X102B06   
1  211050_x_at    -0.60609845997949   -0.71211732736651   -0.24902371657542   
2    220399_at    0.297761884637424   0.108554028539094   0.529915688630775   
3  202115_s_at  -0.0207199586993729  -0.329812842420013  -0.455559869614984   
4    224501_at   0.0191163830257484  -0.508559867759712  0.0562016961050977   

                   4                   5                    6    \
0              X103B41             X104B91              X105B13   
1     0.47339419538929   -0.54907184258026     -0.3092005016715   
2  -0.0236977596184555  -0.240674552924416   -0.186094550508416   
3    0.335866503578156  0.0914735642808466  -0.0649335936114834   
4  -0.0112785620922615   0.189045651599998    0.696562269810068   

                   7                    8                   9    ...  \
0              X106B55               X10B88             X110B34  ...   
1     0.41897323719004     -0.3946572303695   -0.47251557515557  ...   
2  0.00280455771169397   -0.131805918000736   0.113294687596224  ...   
3    0.232192293522567  -0.0245333551689031   0.390957101123317  ...   
4    0.192981870643838   -0.319990873927932  -0.243858578920162  ...   

                 242                  243                 244  \
0             X85A03               X86A40              X87A79   
1   0.19425738895826  -0.0147835375762102    0.13407539167209   
2  0.158997493199424    0.228584615788685   -1.61824983812591   
3  0.611860013300097    0.303447871810647  0.0735930672214167   
4  0.209863932679728    0.215091978389048  -0.323448141350642   

                   245                246                 247  \
0               X88A67              X8B87              X90A63   
1    -0.19990245782577   0.11891965170859   -0.37927029343351   
2   -0.237390400226426  0.228422630263264  -0.433716710549096   
3   -0.082834904637803  0.133630190202147  0.0553951474851369   
4  -0.0203078647708521  0.214314624350358   0.212822716079898   

                  248                  249                  250  \
0              X94A16               X96A21               X99A50   
1    0.82007590513398  -0.0455767367670501  -0.0245526540187702   
2   -1.76092467281198    0.290774879921944    0.248706789784914   
3  0.0711540480339563   -0.114437082375583    0.242832779585116   
4   0.368002424465348    0.207985730041758    0.267992659684308   

                   251  
0                X9B52  
1     0.27066630092569  
2  -0.0227919448987057  
3    0.218242406811207  
4    0.791908891893848  

[5 rows x 252 columns]

In [4]:
#Making a copy of the original file
df_main = df_main.T
df_main.columns = df_main.iloc[0] 
df_main = df_main[1:]
df = df_main.copy()

In [5]:
df.head()

0    probe        211050_x_at            220399_at          202115_s_at  \
1  X100B08  -0.60609845997949    0.297761884637424  -0.0207199586993729   
2  X101B88  -0.71211732736651    0.108554028539094   -0.329812842420013   
3  X102B06  -0.24902371657542    0.529915688630775   -0.455559869614984   
4  X103B41   0.47339419538929  -0.0236977596184555    0.335866503578156   
5  X104B91  -0.54907184258026   -0.240674552924416   0.0914735642808466   

0            224501_at         227347_x_at         205483_s_at  \
1   0.0191163830257484   0.217003285274825  -0.959765619734105   
2   -0.508559867759712   0.750268555716166  -0.739373617253365   
3   0.0562016961050977   0.469504094265635   0.436695502974866   
4  -0.0112785620922615  -0.182405554283685   0.913890761259296   
5    0.189045651599998  -0.472259746631345    1.18484495173342   

0          212283_at        212285_s_at         217419_x_at  ... 216922_x_at  \
1   0.31905965380238   0.26743053282687  0.0546245393746343  ...    0.585269   
2  -0.63507596067365  0.404148351178069   0.553476301226864  ...     0.34349   
3  0.253222377924609  0.236893737236249   0.263843536673264  ...    -1.19229   
4  0.134584773097739   0.46841163348552   0.508556696455114  ...    0.233813   
5  0.204097424289389  0.397235706101899   0.107319891585904  ...  -0.0637161   

0  208331_at 211461_at 232402_at 207647_at  207646_s_at  207916_at  \
1   0.677886  0.294886  0.334739 -0.922423     0.525822    0.65593   
2   -0.37717 -0.146836 -0.027103  0.108189    -0.108456   0.800626   
3  -0.264924 -0.403073 -0.273291  0.408303     0.256424  0.0524039   
4  -0.182134 -0.132613  0.314508  0.147834    -0.766659   -1.51796   
5 -0.0511183 -0.149773 -0.304061   -1.2299 -0.000545084   0.574513   

0 202829_s_at 214950_at 208164_s_at  
1   -0.373383  0.533709      0.9551  
2   0.0473873 -0.630866    0.550008  
3  -0.0767656  0.170939   -0.016254  
4    0.487872  0.394624    0.440938  
5    0.597032  0.196972    0.120323  

[5 rows x 24712 columns]

### Phenotype - Output variables (Target Variables)

In [6]:
df_output = pd.read_csv("miller2005_public_clinicalMatrix", sep='\t', index_col=0)

In [7]:
df_output = df_output.reset_index(drop=True)
df1 = df_output.copy()
df1.head()

AgeAtDiagnosis DLDAerror              DSS  DSS.time ERstatus  \
0              68        No  AliveOrCensored      12.0      ER+   
1              40        No  AliveOrCensored      12.0      ER-   
2              51        No  AliveOrCensored      12.0      ER+   
3              52       Yes              NaN       NaN      ER-   
4              80       Yes  AliveOrCensored       4.0      ER+   

  EistonHistologicGrade LymphNodeStatus PgRstatus  TumorSize _INTEGRATION  \
0                    G1             LN-      PgR-        9.0      X100B08   
1                    G3             LN-      PgR-       12.0      X101B88   
2                    G3             LN-      PgR+       26.0      X102B06   
3                    G3             LN+      PgR-       22.0      X103B41   
4                    G3             NaN      PgR+       24.0      X104B91   

  _PATIENT p53DLDAclassResult p53mutant _GENOMIC_ID_miller2005  
0  X100B08            WT-like      p53-                X100B08  
1  X101B88            MT-like      p53+                X101B88  
2  X102B06            MT-like      p53+                X102B06  
3  X103B41            MT-like      p53-                X103B41  
4  X104B91            WT-like      p53+                X104B91

In [8]:
df1.columns

Index(['AgeAtDiagnosis', 'DLDAerror', 'DSS', 'DSS.time', 'ERstatus',
       'EistonHistologicGrade', 'LymphNodeStatus', 'PgRstatus', 'TumorSize',
       '_INTEGRATION', '_PATIENT', 'p53DLDAclassResult', 'p53mutant',
       '_GENOMIC_ID_miller2005'],
      dtype='object')

In [9]:
print(df1.isnull().sum())
plt.figure(figsize=(25,5))
plt.bar(df1.columns, df1.isnull().sum())

AgeAtDiagnosis             0
DLDAerror                  0
DSS                       15
DSS.time                  15
ERstatus                   4
EistonHistologicGrade      2
LymphNodeStatus            9
PgRstatus                  0
TumorSize                  0
_INTEGRATION               0
_PATIENT                   0
p53DLDAclassResult         0
p53mutant                  0
_GENOMIC_ID_miller2005     0
dtype: int64


<BarContainer object of 14 artists>

# 1. Target Variable - DLDAerror:

### Combining both Input data and Output data(DLDAerror)

Merging all the input data from the dataframe "df" and "DLDAerror" variable from the dataframe "df1"


In [10]:
df_final = pd.merge(df,df1[['DLDAerror']], left_on=['probe'],right_on=[df1['_PATIENT']],how='inner')

In [11]:
df_final.head()

probe        211050_x_at            220399_at          202115_s_at  \
0  X100B08  -0.60609845997949    0.297761884637424  -0.0207199586993729   
1  X101B88  -0.71211732736651    0.108554028539094   -0.329812842420013   
2  X102B06  -0.24902371657542    0.529915688630775   -0.455559869614984   
3  X103B41   0.47339419538929  -0.0236977596184555    0.335866503578156   
4  X104B91  -0.54907184258026   -0.240674552924416   0.0914735642808466   

             224501_at         227347_x_at         205483_s_at  \
0   0.0191163830257484   0.217003285274825  -0.959765619734105   
1   -0.508559867759712   0.750268555716166  -0.739373617253365   
2   0.0562016961050977   0.469504094265635   0.436695502974866   
3  -0.0112785620922615  -0.182405554283685   0.913890761259296   
4    0.189045651599998  -0.472259746631345    1.18484495173342   

           212283_at        212285_s_at         217419_x_at  ...  208331_at  \
0   0.31905965380238   0.26743053282687  0.0546245393746343  ...   0.677886   
1  -0.63507596067365  0.404148351178069   0.553476301226864  ...   -0.37717   
2  0.253222377924609  0.236893737236249   0.263843536673264  ...  -0.264924   
3  0.134584773097739   0.46841163348552   0.508556696455114  ...  -0.182134   
4  0.204097424289389  0.397235706101899   0.107319891585904  ... -0.0511183   

  211461_at 232402_at 207647_at  207646_s_at  207916_at 202829_s_at 214950_at  \
0  0.294886  0.334739 -0.922423     0.525822    0.65593   -0.373383  0.533709   
1 -0.146836 -0.027103  0.108189    -0.108456   0.800626   0.0473873 -0.630866   
2 -0.403073 -0.273291  0.408303     0.256424  0.0524039  -0.0767656  0.170939   
3 -0.132613  0.314508  0.147834    -0.766659   -1.51796    0.487872  0.394624   
4 -0.149773 -0.304061   -1.2299 -0.000545084   0.574513    0.597032  0.196972   

  208164_s_at DLDAerror  
0      0.9551        No  
1    0.550008        No  
2   -0.016254        No  
3    0.440938       Yes  
4    0.120323       Yes  

[5 rows x 24713 columns]

## Imbalance data in the Target Variable

In [12]:
print(df_final['DLDAerror'].value_counts())
plt.figure(figsize=(1.5,1.5), dpi=227)
plt.pie(df_final['DLDAerror'].value_counts(), labels=['No', 'Yes'], textprops={'fontsize':6}, startangle=0, autopct='%1.1f%%')
plt.show

No     213
Yes     38
Name: DLDAerror, dtype: int64


<function matplotlib.pyplot.show(close=None, block=None)>

We could see that the Target Variable has 

No - 213  &  Yes - 38,

Data looks so imbalanced. Imbalanced dataset may make the model to perform poor.

In [13]:
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import RFE
from sklearn import preprocessing


X = df_final.drop(columns = ['probe','DLDAerror'],axis=1)
Y = df_final['DLDAerror']

X=X.astype(float)
le = preprocessing.LabelEncoder()
Y = le.fit_transform(Y)


Label Encoder is used to encode the categorical value into numerical value in the Target Variable

In [14]:
print(X.shape,Y.shape)

(251, 24711) (251,)


We could see that the X shape is 251 rows and 24711 columns.
More columns can make the model to perform poor, so its better to reject the features, which is highly correlated. 

Here, we have choosen Forward Feature selection to select the best feature for the target variable - DLDAerror.

### Feature Selection - Forward Feature Selection:

In [15]:
from sklearn.feature_selection import f_regression
ffs = f_regression(X,Y)

variable = [ ]
for i in range(0,len(X.columns)-1):
    if ffs[0][i] >=10:
       variable.append(X.columns[i])

In [16]:
df_best = df_final[variable]

df_best.shape

(251, 2396)

We have choosen f-value as 10. Based on the value, 2396 features were selected out of 24711.

In [17]:
#converting all the datatypes of dataframe to float
df_best = df_best.astype(float)
df_best.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 251 entries, 0 to 250
Columns: 2396 entries, 205483_s_at to 206779_s_at
dtypes: float64(2396)
memory usage: 4.6 MB


## Principle component analysis & Decision Tree:

Principle Component Analysis is used to reduce the dimension of the obtained feature(2396).

PCA reduce the total features less than or equal to the samples.

In [18]:
#Decision Tree model is used here for the classification

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
from sklearn.decomposition import PCA

RF_acc = []
val = []
acc =0
for i in range(5,250,5):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
   
  X_train1, X_test1, y_train1, y_test1 = train_test_split(x_pca, Y, train_size=0.7, random_state=11, shuffle=True)

  dtree = DecisionTreeClassifier()
  dtree.fit(X_train1, y_train1)
  p3 = dtree.predict(X_test1)
  accuracy=accuracy_score(y_test1,p3)*100
  RF_acc.append(accuracy)
  val.append(i)
  if (acc < accuracy):
    acc = accuracy
    best_feat = i
    best_pca = x_pca

## Visualising and selecting the best number of features

In [19]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,RF_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

C:\Users\kvman\anaconda3\lib\site-packages\seaborn\_decorators.py:36: FutureWarning: Pass the following variables as keyword args: x, y. From version 0.12, the only valid positional argument will be `data`, and passing other arguments without an explicit keyword will result in an error or misinterpretation.
  warnings.warn(


Maximum Accuracy is  90.78947368421053  Best number of features :  85


## K-Fold cross-Validation

Here, K-Fold cross validation is used for the training and validating the data with the best number of features which we have choosen from the above.

In [20]:
from sklearn.model_selection import KFold
from sklearn.preprocessing import StandardScaler

X_train1, X_test1, y_train1, y_test1 = train_test_split(best_pca, Y, train_size=0.7, random_state=11, shuffle=True)
print(X_train1.shape)
print(X_test1.shape)

acc_score = []
k = 3
kf = KFold(n_splits=k, shuffle=False, random_state=None)
for train_index , test_index in kf.split(X_train1):
  X_train , X_val = X_train1[train_index,:],X_train1[test_index,:]
  y_train , y_val = y_train1[train_index] , y_train1[test_index]
  
  dtree.fit(X_train,y_train)
  pred_values = dtree.predict(X_val)
    
  acc = accuracy_score(pred_values , y_val)
  acc_score.append(acc)
     
avg_acc_score = sum(acc_score)/k
 
print('accuracy of each fold - {}'.format(acc_score))
print('Avg accuracy : {}'.format(avg_acc_score))


pred2=dtree.predict(X_test1)
print("Test Accuracy : ", accuracy_score(y_test1,pred2)*100)

(175, 85)
(76, 85)
accuracy of each fold - [0.847457627118644, 0.8448275862068966, 0.7068965517241379]
Avg accuracy : 0.7997272550165594
Test Accuracy :  85.52631578947368


# 2. Target Variable - PgRstatus:

In [21]:
df = df_main.copy()
df1 =df_output.copy()

### Combining both Input data and Output data(PgRstatus)

Merging all the input data from the dataframe "df" and "PgRstatus" variable from the dataframe "df1"


In [22]:
df_final = pd.merge(df,df1[['PgRstatus']], left_on=['probe'],right_on=[df1['_PATIENT']],how='inner')

In [23]:
df_final.head()

probe        211050_x_at            220399_at          202115_s_at  \
0  X100B08  -0.60609845997949    0.297761884637424  -0.0207199586993729   
1  X101B88  -0.71211732736651    0.108554028539094   -0.329812842420013   
2  X102B06  -0.24902371657542    0.529915688630775   -0.455559869614984   
3  X103B41   0.47339419538929  -0.0236977596184555    0.335866503578156   
4  X104B91  -0.54907184258026   -0.240674552924416   0.0914735642808466   

             224501_at         227347_x_at         205483_s_at  \
0   0.0191163830257484   0.217003285274825  -0.959765619734105   
1   -0.508559867759712   0.750268555716166  -0.739373617253365   
2   0.0562016961050977   0.469504094265635   0.436695502974866   
3  -0.0112785620922615  -0.182405554283685   0.913890761259296   
4    0.189045651599998  -0.472259746631345    1.18484495173342   

           212283_at        212285_s_at         217419_x_at  ...  208331_at  \
0   0.31905965380238   0.26743053282687  0.0546245393746343  ...   0.677886   
1  -0.63507596067365  0.404148351178069   0.553476301226864  ...   -0.37717   
2  0.253222377924609  0.236893737236249   0.263843536673264  ...  -0.264924   
3  0.134584773097739   0.46841163348552   0.508556696455114  ...  -0.182134   
4  0.204097424289389  0.397235706101899   0.107319891585904  ... -0.0511183   

  211461_at 232402_at 207647_at  207646_s_at  207916_at 202829_s_at 214950_at  \
0  0.294886  0.334739 -0.922423     0.525822    0.65593   -0.373383  0.533709   
1 -0.146836 -0.027103  0.108189    -0.108456   0.800626   0.0473873 -0.630866   
2 -0.403073 -0.273291  0.408303     0.256424  0.0524039  -0.0767656  0.170939   
3 -0.132613  0.314508  0.147834    -0.766659   -1.51796    0.487872  0.394624   
4 -0.149773 -0.304061   -1.2299 -0.000545084   0.574513    0.597032  0.196972   

  208164_s_at PgRstatus  
0      0.9551      PgR-  
1    0.550008      PgR-  
2   -0.016254      PgR+  
3    0.440938      PgR-  
4    0.120323      PgR+  

[5 rows x 24713 columns]

## Imbalance data in the Target Variable

In [24]:
print(df_final['PgRstatus'].value_counts())
plt.figure(figsize=(1.5,1.5), dpi=227)
plt.pie(df_final['PgRstatus'].value_counts(), labels=['PgR+', 'PgR-'], textprops={'fontsize':6}, startangle=0, autopct='%1.1f%%')
plt.show

PgR+    190
PgR-     61
Name: PgRstatus, dtype: int64


<function matplotlib.pyplot.show(close=None, block=None)>

We could see that the Target Variable has 

PgR+ : 190  &  PgR- : 61,

Data looks imbalanced. Imbalanced dataset may make the model to perform poor.

In [25]:
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import RFE
from sklearn import preprocessing


X = df_final.drop(columns = ['probe','PgRstatus'],axis=1)
Y = df_final['PgRstatus']

X=X.astype(float)
le = preprocessing.LabelEncoder()
Y = le.fit_transform(Y)


Label Encoder is used to encode the categorical value into numerical value in the Target Variable

In [26]:
from imblearn.over_sampling import SMOTE
from collections import Counter

smote = SMOTE()

# fit predictor and target variable
x_smote, y_smote = smote.fit_resample(X,Y)

print('Original dataset shape', Counter(Y))
print('Resample dataset shape', Counter(y_smote))
x_smote.shape

Original dataset shape Counter({0: 190, 1: 61})
Resample dataset shape Counter({1: 190, 0: 190})


(380, 24711)

Synthetic Minority Oversampling Technique (SMOTE) is used to balance the dataset.

Here, Oversampling is done, since we have less number of samples.

In [27]:
print(x_smote.shape, y_smote.shape)

(380, 24711) (380,)


We could see that the X shape is 380 rows and 24711 columns.
More columns can make the model to perform poor, so its better to reject the features, which is highly correlated. 

Here, we have choosen Forward Feature selection to select the best feature for the target variable - PgRstatus.

### Feature Selection - Forward Feature Selection:

In [28]:
from sklearn.feature_selection import f_regression
ffs = f_regression(x_smote,y_smote)

variable = [ ]
for i in range(0,len(x_smote.columns)-1):
    if ffs[0][i] >=10:
       variable.append(x_smote.columns[i])

In [29]:
df_best = x_smote[variable]

df_best.shape

(380, 6920)

We have choosen f-value as 10. Based on the value, 7623 features were selected out of 24711.

In [30]:
#converting all the datatypes of dataframe to float
df_best = df_best.astype(float)
df_best.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 380 entries, 0 to 379
Columns: 6920 entries, 212283_at to 209772_s_at
dtypes: float64(6920)
memory usage: 20.1 MB


## Principle component analysis, K_Fold & Random Forest:

Principle Component Analysis is used to reduce the dimension of the obtained feature(7623).

PCA reduce the total features less than or equal to the samples.

In [31]:
#Random Forest model is used here for the classification

from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
from sklearn.decomposition import PCA
from sklearn.model_selection import KFold


RF_acc = []
val = []
acc =0
for i in range(5,250,5):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, y_smote, train_size=0.7, random_state=11, shuffle=True)

  #Splitting the training data into training and validation data using K-Fold
  acc_score = []
  k = 3
  kf = KFold(n_splits=k, shuffle=False, random_state=None)
  for train_index , test_index in kf.split(X_train):
    X_train1 , X_test1 = X_train[train_index,:],X_train[test_index,:]
    y_train1 , y_test1 = y_train[train_index] , y_train[test_index]

    RF = RandomForestClassifier(n_estimators=1000)
    RF.fit(X_train1, y_train1)
    pred2=RF.predict(X_test1)
    accuracy = accuracy_score(y_test1,pred2)*100
    acc_score.append(accuracy)
  avg_acc_score = sum(acc_score)/k
  RF_acc.append(avg_acc_score)
  val.append(i)
  if (acc < accuracy):
    acc = accuracy
    best_feat = i
    best_pca = x_pca

Here, K-Fold cross validation is used for the training and validating the data with the best number of features which we have choosen from the above.

This helps the model to avoid overfitting

## Visualising and selecting the best number of features

In [32]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,RF_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

C:\Users\kvman\anaconda3\lib\site-packages\seaborn\_decorators.py:36: FutureWarning: Pass the following variables as keyword args: x, y. From version 0.12, the only valid positional argument will be `data`, and passing other arguments without an explicit keyword will result in an error or misinterpretation.
  warnings.warn(


Maximum Accuracy is  97.72727272727273  Best number of features :  205


## GridSearchCV

GridSearchCV is used to tuning the parameters of the model - Random Forest.

In [33]:
#GridSearchCV
from sklearn.model_selection import GridSearchCV

X_train1, X_test1, y_train1, y_test1 = train_test_split(best_pca, y_smote, train_size=0.7, random_state=11, shuffle=True)

parameters = {'max_depth': [5,10,20],
 'max_features': ['auto', 'sqrt'],
 'n_estimators': [200,500,1000]}

clf = GridSearchCV(RF, parameters)
clf.fit(X_train1, y_train1)

GridSearchCV(estimator=RandomForestClassifier(n_estimators=1000),
             param_grid={'max_depth': [5, 10, 20],
                         'max_features': ['auto', 'sqrt'],
                         'n_estimators': [200, 500, 1000]})

In [34]:
clf.best_params_

{'max_depth': 5, 'max_features': 'auto', 'n_estimators': 200}

Best Parameters for the Random Forest is identified using gridsearchCV

## Evaluation on Test Data

In [35]:
X_train1, X_test1, y_train1, y_test1 = train_test_split(best_pca, y_smote, train_size=0.7, random_state=11, shuffle=True)
print(X_train1.shape)
print(X_test1.shape)

acc_score = []
k = 3
RF = RandomForestClassifier(max_depth = 5, n_estimators=500, max_features='sqrt')
kf = KFold(n_splits=k, shuffle=False, random_state=None)
for train_index , test_index in kf.split(X_train1):
  X_train , X_val = X_train1[train_index,:],X_train1[test_index,:]
  y_train , y_val = y_train1[train_index] , y_train1[test_index]
  
  RF.fit(X_train,y_train)
  pred_values = RF.predict(X_val)
    
  acc = accuracy_score(pred_values , y_val)
  acc_score.append(acc)
     
avg_acc_score = sum(acc_score)/k
 
print('accuracy of each fold - {}'.format(acc_score))
print('Avg accuracy : {}'.format(avg_acc_score))


pred2=RF.predict(X_test1)
print("Test Accuracy : ", accuracy_score(y_test1,pred2)*100)

(266, 205)
(114, 205)
accuracy of each fold - [0.9438202247191011, 0.9550561797752809, 0.9772727272727273]
Avg accuracy : 0.958716377255703
Test Accuracy :  98.24561403508771


Conclusion:

Model Performs good in both Training and Testing data.

Required Number of features to achieve good accuracy = **230**

Training Accuracy =**96%**

Testing Accuracy = **94%**

# 3. Target Variable - p53DLDAclassResult:

In [36]:
df = df_main.copy()
df1 =df_output.copy()

### Combining both Input data and Output data(p53DLDAclassResult)

Merging all the input data from the dataframe "df" and "p53DLDAclassResult" variable from the dataframe "df1"


In [37]:
df_final = pd.merge(df,df1[['p53DLDAclassResult']], left_on=['probe'],right_on=[df1['_PATIENT']],how='inner')

In [38]:
df_final.head()

probe        211050_x_at            220399_at          202115_s_at  \
0  X100B08  -0.60609845997949    0.297761884637424  -0.0207199586993729   
1  X101B88  -0.71211732736651    0.108554028539094   -0.329812842420013   
2  X102B06  -0.24902371657542    0.529915688630775   -0.455559869614984   
3  X103B41   0.47339419538929  -0.0236977596184555    0.335866503578156   
4  X104B91  -0.54907184258026   -0.240674552924416   0.0914735642808466   

             224501_at         227347_x_at         205483_s_at  \
0   0.0191163830257484   0.217003285274825  -0.959765619734105   
1   -0.508559867759712   0.750268555716166  -0.739373617253365   
2   0.0562016961050977   0.469504094265635   0.436695502974866   
3  -0.0112785620922615  -0.182405554283685   0.913890761259296   
4    0.189045651599998  -0.472259746631345    1.18484495173342   

           212283_at        212285_s_at         217419_x_at  ...  208331_at  \
0   0.31905965380238   0.26743053282687  0.0546245393746343  ...   0.677886   
1  -0.63507596067365  0.404148351178069   0.553476301226864  ...   -0.37717   
2  0.253222377924609  0.236893737236249   0.263843536673264  ...  -0.264924   
3  0.134584773097739   0.46841163348552   0.508556696455114  ...  -0.182134   
4  0.204097424289389  0.397235706101899   0.107319891585904  ... -0.0511183   

  211461_at 232402_at 207647_at  207646_s_at  207916_at 202829_s_at 214950_at  \
0  0.294886  0.334739 -0.922423     0.525822    0.65593   -0.373383  0.533709   
1 -0.146836 -0.027103  0.108189    -0.108456   0.800626   0.0473873 -0.630866   
2 -0.403073 -0.273291  0.408303     0.256424  0.0524039  -0.0767656  0.170939   
3 -0.132613  0.314508  0.147834    -0.766659   -1.51796    0.487872  0.394624   
4 -0.149773 -0.304061   -1.2299 -0.000545084   0.574513    0.597032  0.196972   

  208164_s_at p53DLDAclassResult  
0      0.9551            WT-like  
1    0.550008            MT-like  
2   -0.016254            MT-like  
3    0.440938            MT-like  
4    0.120323            WT-like  

[5 rows x 24713 columns]

## Imbalance data in the Target Variable

In [39]:
print(df_final['p53DLDAclassResult'].value_counts())
plt.figure(figsize=(1.5,1.5), dpi=227)
plt.pie(df_final['p53DLDAclassResult'].value_counts(), labels=['WT-like', 'MT-like'], textprops={'fontsize':6}, startangle=0, autopct='%1.1f%%')
plt.show

WT-like    179
MT-like     72
Name: p53DLDAclassResult, dtype: int64


<function matplotlib.pyplot.show(close=None, block=None)>

We could see that the Target Variable has 

WT-like : 179  &  MT-like : 72,

Data looks imbalanced. Imbalanced dataset may make the model to perform poor.

In [40]:
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import RFE
from sklearn import preprocessing


X = df_final.drop(columns = ['probe','p53DLDAclassResult'],axis=1)
Y = df_final['p53DLDAclassResult']

X=X.astype(float)
le = preprocessing.LabelEncoder()
Y = le.fit_transform(Y)


Label Encoder is used to encode the categorical value into numerical value in the Target Variable

In [41]:
from imblearn.over_sampling import SMOTE
from collections import Counter

smote = SMOTE()

# fit predictor and target variable
x_smote, y_smote = smote.fit_resample(X,Y)

print('Original dataset shape', Counter(Y))
print('Resample dataset shape', Counter(y_smote))
x_smote.shape

Original dataset shape Counter({1: 179, 0: 72})
Resample dataset shape Counter({1: 179, 0: 179})


(358, 24711)

Synthetic Minority Oversampling Technique (SMOTE) is used to balance the dataset.

Here, Oversampling is done, since we have less number of samples.

In [42]:
print(x_smote.shape, y_smote.shape)

(358, 24711) (358,)


We could see that the X shape is 380 rows and 24711 columns.
More columns can make the model to perform poor, so its better to reject the features, which is highly correlated. 

Here, we have choosen Forward Feature selection to select the best feature for the target variable - p53DLDAclassResult.

### Feature Selection - Forward Feature Selection:

In [43]:
from sklearn.feature_selection import f_regression
ffs = f_regression(x_smote,y_smote)

variable = [ ]
for i in range(0,len(x_smote.columns)-1):
    if ffs[0][i] >=10:
       variable.append(x_smote.columns[i])

In [44]:
df_best = x_smote[variable]

df_best.shape

(358, 10346)

We have choosen f-value as 10. Based on the value, 10846 features were selected out of 24711.

In [45]:
#converting all the datatypes of dataframe to float
df_best = df_best.astype(float)
df_best.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 358 entries, 0 to 357
Columns: 10346 entries, 220399_at to 202829_s_at
dtypes: float64(10346)
memory usage: 28.3 MB


## Principle component analysis & Logistic Regression:

Principle Component Analysis is used to reduce the dimension of the obtained feature(10846).

PCA reduce the total features less than or equal to the samples.

In [46]:
from re import A
#Logistic Regression model is used here for the classification
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
from sklearn.decomposition import PCA

lr_acc = []
val = []
acc =0
for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, y_smote, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  #Splitting the training data into training and validation data using K-Fold
  acc_score = []
  k = 3
  kf = KFold(n_splits=k, shuffle=False, random_state=None)
  for train_index , test_index in kf.split(X_train):
    X_train1 , X_test1 = X_train[train_index,:],X_train[test_index,:]
    y_train1 , y_test1 = y_train[train_index] , y_train[test_index]

    logreg = LogisticRegression(max_iter=500)
    logreg.fit(X_train1, y_train1)
    pred1 = logreg.predict(X_test1)
    accuracy=accuracy_score(y_test1,pred1)*100
    acc_score.append(accuracy)
     
  avg_acc_score = sum(acc_score)/k
  lr_acc.append(avg_acc_score)
  val.append(i)
  if (acc < avg_acc_score):
    acc = avg_acc_score
    best_feat = i
    best_pca = x_pca

Here, K-Fold cross validation is used for the training and validating the data with the best number of features which we have choosen from the above.

This helps the model to avoid overfitting

## Visualising and selecting the best number of features

In [47]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,lr_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

C:\Users\kvman\anaconda3\lib\site-packages\seaborn\_decorators.py:36: FutureWarning: Pass the following variables as keyword args: x, y. From version 0.12, the only valid positional argument will be `data`, and passing other arguments without an explicit keyword will result in an error or misinterpretation.
  warnings.warn(


Maximum Accuracy is  98.40791738382099  Best number of features :  60


### Predicting the Test Data

In [48]:
accuracy_score(y_test,logreg.predict(X_test))*100

96.29629629629629

Conclusion:

Model Performs good in both Training and Testing data.

Required Number of features to achieve good accuracy = **90**

Training Accuracy =**98%**

Testing Accuracy = **95%**

# 4. Target Variable - p53mutant:

In [49]:
df = df_main.copy()
df1 =df_output.copy()

### Combining both Input data and Output data(p53mutant)

Merging all the input data from the dataframe "df" and "p53mutant" variable from the dataframe "df1"


In [50]:
df_final = pd.merge(df,df1[['p53mutant']], left_on=['probe'],right_on=[df1['_PATIENT']],how='inner')

In [51]:
df_final.head()

probe        211050_x_at            220399_at          202115_s_at  \
0  X100B08  -0.60609845997949    0.297761884637424  -0.0207199586993729   
1  X101B88  -0.71211732736651    0.108554028539094   -0.329812842420013   
2  X102B06  -0.24902371657542    0.529915688630775   -0.455559869614984   
3  X103B41   0.47339419538929  -0.0236977596184555    0.335866503578156   
4  X104B91  -0.54907184258026   -0.240674552924416   0.0914735642808466   

             224501_at         227347_x_at         205483_s_at  \
0   0.0191163830257484   0.217003285274825  -0.959765619734105   
1   -0.508559867759712   0.750268555716166  -0.739373617253365   
2   0.0562016961050977   0.469504094265635   0.436695502974866   
3  -0.0112785620922615  -0.182405554283685   0.913890761259296   
4    0.189045651599998  -0.472259746631345    1.18484495173342   

           212283_at        212285_s_at         217419_x_at  ...  208331_at  \
0   0.31905965380238   0.26743053282687  0.0546245393746343  ...   0.677886   
1  -0.63507596067365  0.404148351178069   0.553476301226864  ...   -0.37717   
2  0.253222377924609  0.236893737236249   0.263843536673264  ...  -0.264924   
3  0.134584773097739   0.46841163348552   0.508556696455114  ...  -0.182134   
4  0.204097424289389  0.397235706101899   0.107319891585904  ... -0.0511183   

  211461_at 232402_at 207647_at  207646_s_at  207916_at 202829_s_at 214950_at  \
0  0.294886  0.334739 -0.922423     0.525822    0.65593   -0.373383  0.533709   
1 -0.146836 -0.027103  0.108189    -0.108456   0.800626   0.0473873 -0.630866   
2 -0.403073 -0.273291  0.408303     0.256424  0.0524039  -0.0767656  0.170939   
3 -0.132613  0.314508  0.147834    -0.766659   -1.51796    0.487872  0.394624   
4 -0.149773 -0.304061   -1.2299 -0.000545084   0.574513    0.597032  0.196972   

  208164_s_at p53mutant  
0      0.9551      p53-  
1    0.550008      p53+  
2   -0.016254      p53+  
3    0.440938      p53-  
4    0.120323      p53+  

[5 rows x 24713 columns]

## Imbalance data in the Target Variable

In [52]:
print(df_final['p53mutant'].value_counts())
plt.figure(figsize=(1.5,1.5), dpi=227)
plt.pie(df_final['p53mutant'].value_counts(), labels=['P53-', 'P53+'], textprops={'fontsize':6}, startangle=0, autopct='%1.1f%%')
plt.show

p53-    193
p53+     58
Name: p53mutant, dtype: int64


<function matplotlib.pyplot.show(close=None, block=None)>

We could see that the Target Variable has 

P53- : 193  &  P53+ : 58,

Data looks imbalanced. Imbalanced dataset may make the model to perform poor.

In [53]:
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import RFE
from sklearn import preprocessing


X = df_final.drop(columns = ['probe','p53mutant'],axis=1)
Y = df_final['p53mutant']

X=X.astype(float)
le = preprocessing.LabelEncoder()
Y = le.fit_transform(Y)


Label Encoder is used to encode the categorical value into numerical value in the Target Variable

In [54]:
from imblearn.over_sampling import SMOTE
from collections import Counter

smote = SMOTE()

# fit predictor and target variable
x_smote, y_smote = smote.fit_resample(X,Y)

print('Original dataset shape', Counter(Y))
print('Resample dataset shape', Counter(y_smote))
x_smote.shape

Original dataset shape Counter({1: 193, 0: 58})
Resample dataset shape Counter({1: 193, 0: 193})


(386, 24711)

Synthetic Minority Oversampling Technique (SMOTE) is used to balance the dataset.

Here, Oversampling is done, since we have less number of samples.

In [55]:
print(x_smote.shape, y_smote.shape)

(386, 24711) (386,)


We could see that the X shape is 386 rows and 24711 columns.
More columns can make the model to perform poor, so its better to reject the features, which is highly correlated. 

Here, we have choosen Forward Feature selection to select the best feature for the target variable - p53mutant.

### Feature Selection - Forward Feature Selection:

In [56]:
from sklearn.feature_selection import f_regression
ffs = f_regression(x_smote,y_smote)

variable = [ ]
for i in range(0,len(x_smote.columns)-1):
    if ffs[0][i] >=10:
       variable.append(x_smote.columns[i])

In [57]:
df_best = x_smote[variable]

df_best.shape

(386, 8702)

We have choosen f-value as 10. Based on the value, 9215 features were selected out of 24711.

In [58]:
#converting all the datatypes of dataframe to float
df_best = df_best.astype(float)
df_best.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 386 entries, 0 to 385
Columns: 8702 entries, 220399_at to 202829_s_at
dtypes: float64(8702)
memory usage: 25.6 MB


## Principle component analysis & SVC:

Principle Component Analysis is used to reduce the dimension of the obtained feature(9215).

PCA reduce the total features less than or equal to the samples.

In [59]:
from re import A
#Support Vector Classification (SVC) model is used here for the classification
from sklearn import svm
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
from sklearn.decomposition import PCA

svr_acc = []
val = []
acc =0
for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, y_smote, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  #Splitting the training data into training and validation data using K-Fold
  acc_score = []
  k = 3
  kf = KFold(n_splits=k, shuffle=False, random_state=None)
  for train_index , test_index in kf.split(X_train):
    X_train1 , X_test1 = X_train[train_index,:],X_train[test_index,:]
    y_train1 , y_test1 = y_train[train_index] , y_train[test_index]

    svc = svm.SVC(kernel='rbf', C =10)
    svc.fit(X_train1,y_train1)
    pred3 = svc.predict(X_test1)
    accuracy = accuracy_score(y_test1,pred3)*100
    acc_score.append(accuracy)
     
  avg_acc_score = sum(acc_score)/k
  svr_acc.append(avg_acc_score)
  val.append(i)
  if (acc < avg_acc_score):
    acc = avg_acc_score
    best_feat = i
    best_pca = x_pca

Here, K-Fold cross validation is used for the training and validating the data with the best number of features which we have choosen from the above.

This helps the model to avoid overfitting

## Visualising and selecting the best number of features

In [60]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,svr_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

C:\Users\kvman\anaconda3\lib\site-packages\seaborn\_decorators.py:36: FutureWarning: Pass the following variables as keyword args: x, y. From version 0.12, the only valid positional argument will be `data`, and passing other arguments without an explicit keyword will result in an error or misinterpretation.
  warnings.warn(


Maximum Accuracy is  94.44444444444444  Best number of features :  60


### Predicting the Test Data

In [61]:
accuracy_score(y_test,svc.predict(X_test))*100

88.79310344827587

Conclusion:

Model Performs good in both Training and Testing data.

Required Number of features to achieve good accuracy = **70**

Training Accuracy =**93%**

Testing Accuracy = **91%**

# 5. Target Variable - LymphNodeStatus:

In [62]:
df = df_main.copy()
df1 =df_output.copy()

### Combining both Input data and Output data(LymphNodeStatus)

Merging all the input data from the dataframe "df" and "LymphNodeStatus" variable from the dataframe "df1"


In [63]:
df_final = pd.merge(df,df1[['LymphNodeStatus']], left_on=['probe'],right_on=[df1['_PATIENT']],how='inner')

In [64]:
df_final.head()

probe        211050_x_at            220399_at          202115_s_at  \
0  X100B08  -0.60609845997949    0.297761884637424  -0.0207199586993729   
1  X101B88  -0.71211732736651    0.108554028539094   -0.329812842420013   
2  X102B06  -0.24902371657542    0.529915688630775   -0.455559869614984   
3  X103B41   0.47339419538929  -0.0236977596184555    0.335866503578156   
4  X104B91  -0.54907184258026   -0.240674552924416   0.0914735642808466   

             224501_at         227347_x_at         205483_s_at  \
0   0.0191163830257484   0.217003285274825  -0.959765619734105   
1   -0.508559867759712   0.750268555716166  -0.739373617253365   
2   0.0562016961050977   0.469504094265635   0.436695502974866   
3  -0.0112785620922615  -0.182405554283685   0.913890761259296   
4    0.189045651599998  -0.472259746631345    1.18484495173342   

           212283_at        212285_s_at         217419_x_at  ...  208331_at  \
0   0.31905965380238   0.26743053282687  0.0546245393746343  ...   0.677886   
1  -0.63507596067365  0.404148351178069   0.553476301226864  ...   -0.37717   
2  0.253222377924609  0.236893737236249   0.263843536673264  ...  -0.264924   
3  0.134584773097739   0.46841163348552   0.508556696455114  ...  -0.182134   
4  0.204097424289389  0.397235706101899   0.107319891585904  ... -0.0511183   

  211461_at 232402_at 207647_at  207646_s_at  207916_at 202829_s_at 214950_at  \
0  0.294886  0.334739 -0.922423     0.525822    0.65593   -0.373383  0.533709   
1 -0.146836 -0.027103  0.108189    -0.108456   0.800626   0.0473873 -0.630866   
2 -0.403073 -0.273291  0.408303     0.256424  0.0524039  -0.0767656  0.170939   
3 -0.132613  0.314508  0.147834    -0.766659   -1.51796    0.487872  0.394624   
4 -0.149773 -0.304061   -1.2299 -0.000545084   0.574513    0.597032  0.196972   

  208164_s_at LymphNodeStatus  
0      0.9551             LN-  
1    0.550008             LN-  
2   -0.016254             LN-  
3    0.440938             LN+  
4    0.120323             NaN  

[5 rows x 24713 columns]

## Missing Values

Removing the missing values

In [65]:
df_final.isna().sum()

probe              0
211050_x_at        0
220399_at          0
202115_s_at        0
224501_at          0
                  ..
207916_at          0
202829_s_at        0
214950_at          0
208164_s_at        0
LymphNodeStatus    9
Length: 24713, dtype: int64

In [66]:
df_final = df_final.dropna()

In [67]:
df_final.isna().sum()

probe              0
211050_x_at        0
220399_at          0
202115_s_at        0
224501_at          0
                  ..
207916_at          0
202829_s_at        0
214950_at          0
208164_s_at        0
LymphNodeStatus    0
Length: 24713, dtype: int64

## Imbalance data in the Target Variable

In [68]:
print(df_final['LymphNodeStatus'].value_counts())
plt.figure(figsize=(1.5,1.5), dpi=227)
plt.pie(df_final['LymphNodeStatus'].value_counts(), labels=['LN-', 'LN+'], textprops={'fontsize':6}, startangle=0, autopct='%1.1f%%')
plt.show

LN-    158
LN+     84
Name: LymphNodeStatus, dtype: int64


<function matplotlib.pyplot.show(close=None, block=None)>

We could see that the Target Variable has 

LN- : 158  &  LN3+ : 84,

Data looks imbalanced. Imbalanced dataset may make the model to perform poor.

In [69]:
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import RFE
from sklearn import preprocessing


X = df_final.drop(columns = ['probe','LymphNodeStatus'],axis=1)
Y = df_final['LymphNodeStatus']

X=X.astype(float)
le = preprocessing.LabelEncoder()
Y = le.fit_transform(Y)


Label Encoder is used to encode the categorical value into numerical value in the Target Variable

In [70]:
from imblearn.over_sampling import SMOTE
from collections import Counter

smote = SMOTE()

# fit predictor and target variable
x_smote, y_smote = smote.fit_resample(X,Y)

print('Original dataset shape', Counter(Y))
print('Resample dataset shape', Counter(y_smote))
x_smote.shape

Original dataset shape Counter({1: 158, 0: 84})
Resample dataset shape Counter({1: 158, 0: 158})


(316, 24711)

Synthetic Minority Oversampling Technique (SMOTE) is used to balance the dataset.

Here, Oversampling is done, since we have less number of samples.

In [71]:
print(x_smote.shape, y_smote.shape)

(316, 24711) (316,)


We could see that the X shape is 316 rows and 24711 columns.
More columns can make the model to perform poor, so its better to reject the features, which is highly correlated. 

Here, we have choosen Forward Feature selection to select the best feature for the target variable - LymphNodeStatus.

### Feature Selection - Forward Feature Selection:

In [72]:
from sklearn.feature_selection import f_regression
ffs = f_regression(x_smote,y_smote)

variable = [ ]
for i in range(0,len(x_smote.columns)-1):
    if ffs[0][i] >=10:
       variable.append(x_smote.columns[i])

In [73]:
df_best = x_smote[variable]

df_best.shape

(316, 1560)

We have choosen f-value as 10. Based on the value, 1882 features were selected out of 24711.

In [74]:
#converting all the datatypes of dataframe to float
df_best = df_best.astype(float)
df_best.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 316 entries, 0 to 315
Columns: 1560 entries, 205483_s_at to 202829_s_at
dtypes: float64(1560)
memory usage: 3.8 MB


## Principle component analysis & Naive Bayes:

Principle Component Analysis is used to reduce the dimension of the obtained feature(9215).

PCA reduce the total features less than or equal to the samples.

In [75]:
#Naive Bayes model is used here for the classification
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
from sklearn.decomposition import PCA
from sklearn.naive_bayes import GaussianNB

nb_acc = []
val = []
acc =0
for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, y_smote, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  #Splitting the training data into training and validation data using K-Fold
  acc_score = []
  k = 3
  kf = KFold(n_splits=k, shuffle=False, random_state=None)
  for train_index , test_index in kf.split(X_train):
    X_train1 , X_test1 = X_train[train_index,:],X_train[test_index,:]
    y_train1 , y_test1 = y_train[train_index] , y_train[test_index]

    GNB = GaussianNB()
    GNB.fit(X_train1, y_train1)
    pred4 = GNB.predict(X_test1)
    accuracy = accuracy_score(y_test1,pred4)*100
    acc_score.append(accuracy)
     
  avg_acc_score = sum(acc_score)/k
  nb_acc.append(avg_acc_score)
  val.append(i)
  if (acc < avg_acc_score):
    acc = avg_acc_score
    best_feat = i
    best_pca = x_pca

Here, K-Fold cross validation is used for the training and validating the data with the best number of features which we have choosen from the above.

This helps the model to avoid overfitting

## Visualising and selecting the best number of features

In [76]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,svr_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

Maximum Accuracy is  87.78847340491177  Best number of features :  220


C:\Users\kvman\anaconda3\lib\site-packages\seaborn\_decorators.py:36: FutureWarning: Pass the following variables as keyword args: x, y. From version 0.12, the only valid positional argument will be `data`, and passing other arguments without an explicit keyword will result in an error or misinterpretation.
  warnings.warn(


### Predicting the Test Data

In [77]:
accuracy_score(y_test,GNB.predict(X_test))*100

86.31578947368422

Conclusion:

Model Performs good in both Training and Testing data.

Required Number of features to achieve good accuracy = **210**

Training Accuracy =**87%**

Testing Accuracy = **88%**

# 6. Target Variable - EistonHistologicGrade:

In [78]:
df = df_main.copy()
df1 =df_output.copy()

### Combining both Input data and Output data(EistonHistologicGrade)

Merging all the input data from the dataframe "df" and "EistonHistologicGrade" variable from the dataframe "df1"


In [79]:
df_final = pd.merge(df,df1[['EistonHistologicGrade']], left_on=['probe'],right_on=[df1['_PATIENT']],how='inner')

In [80]:
df_final.head()

probe        211050_x_at            220399_at          202115_s_at  \
0  X100B08  -0.60609845997949    0.297761884637424  -0.0207199586993729   
1  X101B88  -0.71211732736651    0.108554028539094   -0.329812842420013   
2  X102B06  -0.24902371657542    0.529915688630775   -0.455559869614984   
3  X103B41   0.47339419538929  -0.0236977596184555    0.335866503578156   
4  X104B91  -0.54907184258026   -0.240674552924416   0.0914735642808466   

             224501_at         227347_x_at         205483_s_at  \
0   0.0191163830257484   0.217003285274825  -0.959765619734105   
1   -0.508559867759712   0.750268555716166  -0.739373617253365   
2   0.0562016961050977   0.469504094265635   0.436695502974866   
3  -0.0112785620922615  -0.182405554283685   0.913890761259296   
4    0.189045651599998  -0.472259746631345    1.18484495173342   

           212283_at        212285_s_at         217419_x_at  ...  208331_at  \
0   0.31905965380238   0.26743053282687  0.0546245393746343  ...   0.677886   
1  -0.63507596067365  0.404148351178069   0.553476301226864  ...   -0.37717   
2  0.253222377924609  0.236893737236249   0.263843536673264  ...  -0.264924   
3  0.134584773097739   0.46841163348552   0.508556696455114  ...  -0.182134   
4  0.204097424289389  0.397235706101899   0.107319891585904  ... -0.0511183   

  211461_at 232402_at 207647_at  207646_s_at  207916_at 202829_s_at 214950_at  \
0  0.294886  0.334739 -0.922423     0.525822    0.65593   -0.373383  0.533709   
1 -0.146836 -0.027103  0.108189    -0.108456   0.800626   0.0473873 -0.630866   
2 -0.403073 -0.273291  0.408303     0.256424  0.0524039  -0.0767656  0.170939   
3 -0.132613  0.314508  0.147834    -0.766659   -1.51796    0.487872  0.394624   
4 -0.149773 -0.304061   -1.2299 -0.000545084   0.574513    0.597032  0.196972   

  208164_s_at EistonHistologicGrade  
0      0.9551                    G1  
1    0.550008                    G3  
2   -0.016254                    G3  
3    0.440938                    G3  
4    0.120323                    G3  

[5 rows x 24713 columns]

## Missing Values

Removing the missing values

In [81]:
df_final.isna().sum()

probe                    0
211050_x_at              0
220399_at                0
202115_s_at              0
224501_at                0
                        ..
207916_at                0
202829_s_at              0
214950_at                0
208164_s_at              0
EistonHistologicGrade    2
Length: 24713, dtype: int64

In [82]:
df_final = df_final.dropna()

In [83]:
df_final.isna().sum()

probe                    0
211050_x_at              0
220399_at                0
202115_s_at              0
224501_at                0
                        ..
207916_at                0
202829_s_at              0
214950_at                0
208164_s_at              0
EistonHistologicGrade    0
Length: 24713, dtype: int64

## Imbalance data in the Target Variable

In [84]:
print(df_final['EistonHistologicGrade'].value_counts())
plt.figure(figsize=(1.5,1.5), dpi=227)
plt.pie(df_final['EistonHistologicGrade'].value_counts(), labels=['G2', 'G1', 'G3'], textprops={'fontsize':6}, startangle=0, autopct='%1.1f%%')
plt.show

G2    128
G1     67
G3     54
Name: EistonHistologicGrade, dtype: int64


<function matplotlib.pyplot.show(close=None, block=None)>

We could see that the Target Variable has 

G1 - 67, G2 - 128, G3 - 54

Data looks imbalanced. Imbalanced dataset may make the model to perform poor.

In [85]:
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import RFE
from sklearn import preprocessing


X = df_final.drop(columns = ['probe','EistonHistologicGrade'],axis=1)
Y = df_final['EistonHistologicGrade']

X=X.astype(float)
le = preprocessing.LabelEncoder()
Y = le.fit_transform(Y)


Label Encoder is used to encode the categorical value into numerical value in the Target Variable

In [86]:
from imblearn.over_sampling import SMOTE
from collections import Counter

smote = SMOTE()

# fit predictor and target variable
x_smote, y_smote = smote.fit_resample(X,Y)

print('Original dataset shape', Counter(Y))
print('Resample dataset shape', Counter(y_smote))
x_smote.shape

Original dataset shape Counter({1: 128, 0: 67, 2: 54})
Resample dataset shape Counter({0: 128, 2: 128, 1: 128})


(384, 24711)

Synthetic Minority Oversampling Technique (SMOTE) is used to balance the dataset.

Here, Oversampling is done, since we have less number of samples.

In [87]:
print(x_smote.shape, y_smote.shape)

(384, 24711) (384,)


We could see that the X shape is 384 rows and 24711 columns.
More columns can make the model to perform poor, so its better to reject the features, which is highly correlated. 

Here, we have choosen Forward Feature selection to select the best feature for the target variable - EistonHistologicGrade.

### Feature Selection - Forward Feature Selection:

In [88]:
from sklearn.feature_selection import f_regression
ffs = f_regression(x_smote,y_smote)

variable = [ ]
for i in range(0,len(x_smote.columns)-1):
    if ffs[0][i] >=10:
       variable.append(x_smote.columns[i])

In [89]:
df_best = x_smote[variable]

df_best.shape

(384, 10204)

We have choosen f-value as 10. Based on the value, 10336 features were selected out of 24711.

In [90]:
#converting all the datatypes of dataframe to float
df_best = df_best.astype(float)
df_best.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 384 entries, 0 to 383
Columns: 10204 entries, 202115_s_at to 202829_s_at
dtypes: float64(10204)
memory usage: 29.9 MB


## Principle component analysis & Naive Bayes:

Principle Component Analysis is used to reduce the dimension of the obtained feature(9215).

PCA reduce the total features less than or equal to the samples.

In [91]:
#Naive Bayes model is used here for the classification
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
from sklearn.decomposition import PCA
from sklearn.neighbors import (NeighborhoodComponentsAnalysis, KNeighborsClassifier)
from sklearn.pipeline import Pipeline

nca_acc = []
val = []
acc =0
for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, y_smote, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  #Splitting the training data into training and validation data using K-Fold
  acc_score = []
  k = 3
  kf = KFold(n_splits=k, shuffle=False, random_state=None)
  for train_index , test_index in kf.split(X_train):
    X_train1 , X_test1 = X_train[train_index,:],X_train[test_index,:]
    y_train1 , y_test1 = y_train[train_index] , y_train[test_index]

    

    nca = NeighborhoodComponentsAnalysis(random_state=42)
    knn = KNeighborsClassifier(n_neighbors=3)
    nca_pipe = Pipeline([('nca', nca), ('knn', knn)])
    nca_pipe.fit(X_train1, y_train1)
    accuracy = nca_pipe.score(X_test1, y_test1)*100
    acc_score.append(accuracy)
     
  avg_acc_score = sum(acc_score)/k
  nca_acc.append(avg_acc_score)
  val.append(i)
  if (acc < avg_acc_score):
    acc = avg_acc_score
    best_feat = i
    best_pca = x_pca

Here, K-Fold cross validation is used for the training and validating the data with the best number of features which we have choosen from the above.

This helps the model to avoid overfitting

## Visualising and selecting the best number of features

In [92]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,svr_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

Maximum Accuracy is  77.57802746566792  Best number of features :  140


C:\Users\kvman\anaconda3\lib\site-packages\seaborn\_decorators.py:36: FutureWarning: Pass the following variables as keyword args: x, y. From version 0.12, the only valid positional argument will be `data`, and passing other arguments without an explicit keyword will result in an error or misinterpretation.
  warnings.warn(


### Predicting the Test Data

In [93]:
nca_pipe.score(X_test, y_test)*100

62.93103448275862

## Logistic Regression

In [94]:
#Logistic Regression model is used here for the classification
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
from sklearn.decomposition import PCA

lr_acc = []
val = []
acc =0
for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, y_smote, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  #Splitting the training data into training and validation data using K-Fold
  acc_score = []
  k = 3
  kf = KFold(n_splits=k, shuffle=False, random_state=None)
  for train_index , test_index in kf.split(X_train):
    X_train1 , X_test1 = X_train[train_index,:],X_train[test_index,:]
    y_train1 , y_test1 = y_train[train_index] , y_train[test_index]

    logreg = LogisticRegression(max_iter=500)
    logreg.fit(X_train1, y_train1)
    pred1 = logreg.predict(X_test1)
    accuracy=accuracy_score(y_test1,pred1)*100
    acc_score.append(accuracy)
     
  avg_acc_score = sum(acc_score)/k
  lr_acc.append(avg_acc_score)
  val.append(i)
  if (acc < avg_acc_score):
    acc = avg_acc_score
    best_feat = i
    best_pca = x_pca

C:\Users\kvman\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\kvman\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_

In [95]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,lr_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

Maximum Accuracy is  82.45110278818144  Best number of features :  240


C:\Users\kvman\anaconda3\lib\site-packages\seaborn\_decorators.py:36: FutureWarning: Pass the following variables as keyword args: x, y. From version 0.12, the only valid positional argument will be `data`, and passing other arguments without an explicit keyword will result in an error or misinterpretation.
  warnings.warn(


In [96]:
accuracy_score(y_test,logreg.predict(X_test))*100

69.82758620689656

Conclusion:

Here, we have tried with 2 different models, since first model didn't work as expected.

KNN Model:

Required Number of features to achieve good accuracy = **60**

Training Accuracy =**76%**

Testing Accuracy = **64%**

Logistic Regression Model:

Required Number of features to achieve good accuracy = **160**

Training Accuracy =**79%**

Testing Accuracy = **77%**

# 7. Target Variable - ERstatus:

In [97]:
df = df_main.copy()
df1 =df_output.copy()

### Combining both Input data and Output data(ERstatus)

Merging all the input data from the dataframe "df" and "ERstatus" variable from the dataframe "df1"


In [98]:
df_final = pd.merge(df,df1[['ERstatus']], left_on=['probe'],right_on=[df1['_PATIENT']],how='inner')

In [99]:
df_final.head()

probe        211050_x_at            220399_at          202115_s_at  \
0  X100B08  -0.60609845997949    0.297761884637424  -0.0207199586993729   
1  X101B88  -0.71211732736651    0.108554028539094   -0.329812842420013   
2  X102B06  -0.24902371657542    0.529915688630775   -0.455559869614984   
3  X103B41   0.47339419538929  -0.0236977596184555    0.335866503578156   
4  X104B91  -0.54907184258026   -0.240674552924416   0.0914735642808466   

             224501_at         227347_x_at         205483_s_at  \
0   0.0191163830257484   0.217003285274825  -0.959765619734105   
1   -0.508559867759712   0.750268555716166  -0.739373617253365   
2   0.0562016961050977   0.469504094265635   0.436695502974866   
3  -0.0112785620922615  -0.182405554283685   0.913890761259296   
4    0.189045651599998  -0.472259746631345    1.18484495173342   

           212283_at        212285_s_at         217419_x_at  ...  208331_at  \
0   0.31905965380238   0.26743053282687  0.0546245393746343  ...   0.677886   
1  -0.63507596067365  0.404148351178069   0.553476301226864  ...   -0.37717   
2  0.253222377924609  0.236893737236249   0.263843536673264  ...  -0.264924   
3  0.134584773097739   0.46841163348552   0.508556696455114  ...  -0.182134   
4  0.204097424289389  0.397235706101899   0.107319891585904  ... -0.0511183   

  211461_at 232402_at 207647_at  207646_s_at  207916_at 202829_s_at 214950_at  \
0  0.294886  0.334739 -0.922423     0.525822    0.65593   -0.373383  0.533709   
1 -0.146836 -0.027103  0.108189    -0.108456   0.800626   0.0473873 -0.630866   
2 -0.403073 -0.273291  0.408303     0.256424  0.0524039  -0.0767656  0.170939   
3 -0.132613  0.314508  0.147834    -0.766659   -1.51796    0.487872  0.394624   
4 -0.149773 -0.304061   -1.2299 -0.000545084   0.574513    0.597032  0.196972   

  208164_s_at ERstatus  
0      0.9551      ER+  
1    0.550008      ER-  
2   -0.016254      ER+  
3    0.440938      ER-  
4    0.120323      ER+  

[5 rows x 24713 columns]

## Missing Values

Removing the missing values

In [100]:
df_final.isna().sum()

probe          0
211050_x_at    0
220399_at      0
202115_s_at    0
224501_at      0
              ..
207916_at      0
202829_s_at    0
214950_at      0
208164_s_at    0
ERstatus       4
Length: 24713, dtype: int64

In [101]:
df_final = df_final.dropna()

In [102]:
df_final.isna().sum()

probe          0
211050_x_at    0
220399_at      0
202115_s_at    0
224501_at      0
              ..
207916_at      0
202829_s_at    0
214950_at      0
208164_s_at    0
ERstatus       0
Length: 24713, dtype: int64

## Imbalance data in the Target Variable

In [103]:
print(df_final['ERstatus'].value_counts())
plt.figure(figsize=(1.5,1.5), dpi=227)
plt.pie(df_final['ERstatus'].value_counts(), labels=['ER+', 'ER-'], textprops={'fontsize':6}, startangle=0, autopct='%1.1f%%')
plt.show

ER+    213
ER-     34
Name: ERstatus, dtype: int64


<function matplotlib.pyplot.show(close=None, block=None)>

We could see that the Target Variable has 

ER+ : 213  &  ER- : 34,

Data looks imbalanced. Imbalanced dataset may make the model to perform poor.

In [104]:
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import RFE
from sklearn import preprocessing


X = df_final.drop(columns = ['probe','ERstatus'],axis=1)
Y = df_final['ERstatus']

X=X.astype(float)
le = preprocessing.LabelEncoder()
Y = le.fit_transform(Y)


Label Encoder is used to encode the categorical value into numerical value in the Target Variable

In [105]:
from imblearn.over_sampling import SMOTE
from collections import Counter

smote = SMOTE()

# fit predictor and target variable
x_smote, y_smote = smote.fit_resample(X,Y)

print('Original dataset shape', Counter(Y))
print('Resample dataset shape', Counter(y_smote))
x_smote.shape

Original dataset shape Counter({0: 213, 1: 34})
Resample dataset shape Counter({0: 213, 1: 213})


(426, 24711)

Synthetic Minority Oversampling Technique (SMOTE) is used to balance the dataset.

Here, Oversampling is done, since we have less number of samples.

In [106]:
print(x_smote.shape, y_smote.shape)

(426, 24711) (426,)


We could see that the X shape is 426 rows and 24711 columns.
More columns can make the model to perform poor, so its better to reject the features, which is highly correlated. 

Here, we have choosen Forward Feature selection to select the best feature for the target variable - ERstatus.

### Feature Selection - Forward Feature Selection:

In [107]:
from sklearn.feature_selection import f_regression
ffs = f_regression(x_smote,y_smote)

variable = [ ]
for i in range(0,len(x_smote.columns)-1):
    if ffs[0][i] >=10:
       variable.append(x_smote.columns[i])

In [108]:
df_best = x_smote[variable]

df_best.shape

(426, 10508)

We have choosen f-value as 10. Based on the value, 10502 features were selected out of 24711.

In [109]:
#converting all the datatypes of dataframe to float
df_best = df_best.astype(float)
df_best.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426 entries, 0 to 425
Columns: 10508 entries, 205483_s_at to 207647_at
dtypes: float64(10508)
memory usage: 34.2 MB


## Principle component analysis & Naive Bayes:

Principle Component Analysis is used to reduce the dimension of the obtained feature(10502).

PCA reduce the total features less than or equal to the samples.

In [110]:
#Naive Bayes model is used here for the classification
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
from sklearn.decomposition import PCA
from sklearn.naive_bayes import GaussianNB

nb_acc = []
val = []
acc =0
for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, y_smote, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  #Splitting the training data into training and validation data using K-Fold
  acc_score = []
  k = 3
  kf = KFold(n_splits=k, shuffle=False, random_state=None)
  for train_index , test_index in kf.split(X_train):
    X_train1 , X_test1 = X_train[train_index,:],X_train[test_index,:]
    y_train1 , y_test1 = y_train[train_index] , y_train[test_index]

    GNB = GaussianNB()
    GNB.fit(X_train1, y_train1)
    pred4 = GNB.predict(X_test1)
    accuracy = accuracy_score(y_test1,pred4)*100
    acc_score.append(accuracy)
     
  avg_acc_score = sum(acc_score)/k
  nb_acc.append(avg_acc_score)
  val.append(i)
  if (acc < avg_acc_score):
    acc = avg_acc_score
    best_feat = i
    best_pca = x_pca

Here, K-Fold cross validation is used for the training and validating the data with the best number of features which we have choosen from the above.

This helps the model to avoid overfitting

## Visualising and selecting the best number of features

In [111]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,svr_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

Maximum Accuracy is  97.993265993266  Best number of features :  100


C:\Users\kvman\anaconda3\lib\site-packages\seaborn\_decorators.py:36: FutureWarning: Pass the following variables as keyword args: x, y. From version 0.12, the only valid positional argument will be `data`, and passing other arguments without an explicit keyword will result in an error or misinterpretation.
  warnings.warn(


### Predicting the Test Data

In [112]:
accuracy_score(y_test,GNB.predict(X_test))*100

100.0

Conclusion:

Model Performs good in both Training and Testing data.

Required Number of features to achieve good accuracy = **210**

Training Accuracy =**99%**

Testing Accuracy = **98%**

# 8. Target Variable - DSS:

In [113]:
df = df_main.copy()
df1 =df_output.copy()

### Combining both Input data and Output data(DSS)

Merging all the input data from the dataframe "df" and "DSS" variable from the dataframe "df1"


In [114]:
df_final = pd.merge(df,df1[['DSS']], left_on=['probe'],right_on=[df1['_PATIENT']],how='inner')

In [115]:
df_final.head()

probe        211050_x_at            220399_at          202115_s_at  \
0  X100B08  -0.60609845997949    0.297761884637424  -0.0207199586993729   
1  X101B88  -0.71211732736651    0.108554028539094   -0.329812842420013   
2  X102B06  -0.24902371657542    0.529915688630775   -0.455559869614984   
3  X103B41   0.47339419538929  -0.0236977596184555    0.335866503578156   
4  X104B91  -0.54907184258026   -0.240674552924416   0.0914735642808466   

             224501_at         227347_x_at         205483_s_at  \
0   0.0191163830257484   0.217003285274825  -0.959765619734105   
1   -0.508559867759712   0.750268555716166  -0.739373617253365   
2   0.0562016961050977   0.469504094265635   0.436695502974866   
3  -0.0112785620922615  -0.182405554283685   0.913890761259296   
4    0.189045651599998  -0.472259746631345    1.18484495173342   

           212283_at        212285_s_at         217419_x_at  ...  208331_at  \
0   0.31905965380238   0.26743053282687  0.0546245393746343  ...   0.677886   
1  -0.63507596067365  0.404148351178069   0.553476301226864  ...   -0.37717   
2  0.253222377924609  0.236893737236249   0.263843536673264  ...  -0.264924   
3  0.134584773097739   0.46841163348552   0.508556696455114  ...  -0.182134   
4  0.204097424289389  0.397235706101899   0.107319891585904  ... -0.0511183   

  211461_at 232402_at 207647_at  207646_s_at  207916_at 202829_s_at 214950_at  \
0  0.294886  0.334739 -0.922423     0.525822    0.65593   -0.373383  0.533709   
1 -0.146836 -0.027103  0.108189    -0.108456   0.800626   0.0473873 -0.630866   
2 -0.403073 -0.273291  0.408303     0.256424  0.0524039  -0.0767656  0.170939   
3 -0.132613  0.314508  0.147834    -0.766659   -1.51796    0.487872  0.394624   
4 -0.149773 -0.304061   -1.2299 -0.000545084   0.574513    0.597032  0.196972   

  208164_s_at              DSS  
0      0.9551  AliveOrCensored  
1    0.550008  AliveOrCensored  
2   -0.016254  AliveOrCensored  
3    0.440938              NaN  
4    0.120323  AliveOrCensored  

[5 rows x 24713 columns]

## Missing Values

Removing the missing values

In [116]:
df_final.isna().sum()

probe           0
211050_x_at     0
220399_at       0
202115_s_at     0
224501_at       0
               ..
207916_at       0
202829_s_at     0
214950_at       0
208164_s_at     0
DSS            15
Length: 24713, dtype: int64

In [117]:
df_final['DSS'] = df_final['DSS'].dropna()

In [118]:
df_final.isna().sum()

probe           0
211050_x_at     0
220399_at       0
202115_s_at     0
224501_at       0
               ..
207916_at       0
202829_s_at     0
214950_at       0
208164_s_at     0
DSS            15
Length: 24713, dtype: int64

## Imbalance data in the Target Variable

In [119]:
print(df_final['DSS'].value_counts())
plt.figure(figsize=(1.5,1.5), dpi=227)
plt.pie(df_final['DSS'].value_counts(), labels=['AliveOrCensored', 'DeathFromCancer'], textprops={'fontsize':6}, startangle=0, autopct='%1.1f%%')
plt.show

AliveOrCensored    181
DeathFromCancer     55
Name: DSS, dtype: int64


<function matplotlib.pyplot.show(close=None, block=None)>

We could see that the Target Variable has 

AliveOrCensored    181
DeathFromCancer     55,

Data looks imbalanced. Imbalanced dataset may make the model to perform poor.

In [120]:
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import RFE
from sklearn import preprocessing


X = df_final.drop(columns = ['probe','DSS'],axis=1)
Y = df_final['DSS']

X=X.astype(float)
le = preprocessing.LabelEncoder()
Y = le.fit_transform(Y)


Label Encoder is used to encode the categorical value into numerical value in the Target Variable

In [121]:
from imblearn.over_sampling import SMOTE
from collections import Counter

smote = SMOTE()

# fit predictor and target variable
x_smote, y_smote = smote.fit_resample(X,Y)

print('Original dataset shape', Counter(Y))
print('Resample dataset shape', Counter(y_smote))
x_smote.shape

Original dataset shape Counter({0: 181, 1: 55, 2: 15})
Resample dataset shape Counter({0: 181, 2: 181, 1: 181})


(543, 24711)

Synthetic Minority Oversampling Technique (SMOTE) is used to balance the dataset.

Here, Oversampling is done, since we have less number of samples.

In [122]:
print(x_smote.shape, y_smote.shape)

(543, 24711) (543,)


We could see that the X shape is 543 rows and 24711 columns.
More columns can make the model to perform poor, so its better to reject the features, which is highly correlated. 

Here, we have choosen Forward Feature selection to select the best feature for the target variable - DSS.

### Feature Selection - Forward Feature Selection:

In [123]:
from sklearn.feature_selection import f_regression
ffs = f_regression(x_smote,y_smote)

variable = [ ]
for i in range(0,len(x_smote.columns)-1):
    if ffs[0][i] >=10:
       variable.append(x_smote.columns[i])

In [124]:
df_best = x_smote[variable]

df_best.shape

(543, 5879)

We have choosen f-value as 10. Based on the value, 6580 features were selected out of 24711.

In [125]:
#converting all the datatypes of dataframe to float
df_best = df_best.astype(float)
df_best.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 543 entries, 0 to 542
Columns: 5879 entries, 211050_x_at to 214950_at
dtypes: float64(5879)
memory usage: 24.4 MB


## Principle component analysis & SVC:

Principle Component Analysis is used to reduce the dimension of the obtained feature(6580).

PCA reduce the total features less than or equal to the samples.

In [126]:
from re import A
#Support Vector Classification (SVC) model is used here for the classification
from sklearn import svm
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
from sklearn.decomposition import PCA

svr_acc = []
val = []
acc =0
for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, y_smote, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  #Splitting the training data into training and validation data using K-Fold
  acc_score = []
  k = 3
  kf = KFold(n_splits=k, shuffle=False, random_state=None)
  for train_index , test_index in kf.split(X_train):
    X_train1 , X_test1 = X_train[train_index,:],X_train[test_index,:]
    y_train1 , y_test1 = y_train[train_index] , y_train[test_index]

    svc = svm.SVC(kernel='rbf', C =10)
    svc.fit(X_train1,y_train1)
    pred3 = svc.predict(X_test1)
    accuracy = accuracy_score(y_test1,pred3)*100
    acc_score.append(accuracy)
     
  avg_acc_score = sum(acc_score)/k
  svr_acc.append(avg_acc_score)
  val.append(i)
  if (acc < avg_acc_score):
    acc = avg_acc_score
    best_feat = i
    best_pca = x_pca

Here, K-Fold cross validation is used for the training and validating the data with the best number of features which we have choosen from the above.

This helps the model to avoid overfitting

## Visualising and selecting the best number of features

In [127]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,svr_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

Maximum Accuracy is  97.62737991084447  Best number of features :  80


C:\Users\kvman\anaconda3\lib\site-packages\seaborn\_decorators.py:36: FutureWarning: Pass the following variables as keyword args: x, y. From version 0.12, the only valid positional argument will be `data`, and passing other arguments without an explicit keyword will result in an error or misinterpretation.
  warnings.warn(


### Predicting the Test Data

In [128]:
accuracy_score(y_test,svc.predict(X_test))*100

95.0920245398773

Conclusion:

Model Performs good in both Training and Testing data.

Required Number of features to achieve good accuracy = **70**

Training Accuracy =**96%**

Testing Accuracy = **92%**

# 9. Target Variable - TumorSize:

In [129]:
df = df_main.copy()
df1 =df_output.copy()

### Combining both Input data and Output data(TumorSize)

Merging all the input data from the dataframe "df" and "TumorSize" variable from the dataframe "df1"


In [130]:
df_final = pd.merge(df,df1[['TumorSize']], left_on=['probe'],right_on=[df1['_PATIENT']],how='inner')

In [131]:
df_final.head()

probe        211050_x_at            220399_at          202115_s_at  \
0  X100B08  -0.60609845997949    0.297761884637424  -0.0207199586993729   
1  X101B88  -0.71211732736651    0.108554028539094   -0.329812842420013   
2  X102B06  -0.24902371657542    0.529915688630775   -0.455559869614984   
3  X103B41   0.47339419538929  -0.0236977596184555    0.335866503578156   
4  X104B91  -0.54907184258026   -0.240674552924416   0.0914735642808466   

             224501_at         227347_x_at         205483_s_at  \
0   0.0191163830257484   0.217003285274825  -0.959765619734105   
1   -0.508559867759712   0.750268555716166  -0.739373617253365   
2   0.0562016961050977   0.469504094265635   0.436695502974866   
3  -0.0112785620922615  -0.182405554283685   0.913890761259296   
4    0.189045651599998  -0.472259746631345    1.18484495173342   

           212283_at        212285_s_at         217419_x_at  ...  208331_at  \
0   0.31905965380238   0.26743053282687  0.0546245393746343  ...   0.677886   
1  -0.63507596067365  0.404148351178069   0.553476301226864  ...   -0.37717   
2  0.253222377924609  0.236893737236249   0.263843536673264  ...  -0.264924   
3  0.134584773097739   0.46841163348552   0.508556696455114  ...  -0.182134   
4  0.204097424289389  0.397235706101899   0.107319891585904  ... -0.0511183   

  211461_at 232402_at 207647_at  207646_s_at  207916_at 202829_s_at 214950_at  \
0  0.294886  0.334739 -0.922423     0.525822    0.65593   -0.373383  0.533709   
1 -0.146836 -0.027103  0.108189    -0.108456   0.800626   0.0473873 -0.630866   
2 -0.403073 -0.273291  0.408303     0.256424  0.0524039  -0.0767656  0.170939   
3 -0.132613  0.314508  0.147834    -0.766659   -1.51796    0.487872  0.394624   
4 -0.149773 -0.304061   -1.2299 -0.000545084   0.574513    0.597032  0.196972   

  208164_s_at TumorSize  
0      0.9551       9.0  
1    0.550008      12.0  
2   -0.016254      26.0  
3    0.440938      22.0  
4    0.120323      24.0  

[5 rows x 24713 columns]

## Missing Values

Removing the missing values

In [132]:
df_final.isna().sum()

probe          0
211050_x_at    0
220399_at      0
202115_s_at    0
224501_at      0
              ..
207916_at      0
202829_s_at    0
214950_at      0
208164_s_at    0
TumorSize      0
Length: 24713, dtype: int64

In [133]:
X = df_final.drop(columns = ['probe','TumorSize'],axis=1)
Y = df_final['TumorSize']


In [134]:
print(X.shape, Y.shape)

(251, 24711) (251,)


We could see that the X shape is 251 rows and 24711 columns.
More columns can make the model to perform poor, so its better to reject the features, which is highly correlated. 

Here, we have choosen Forward Feature selection to select the best feature for the target variable - TumorSize.

### Feature Selection - Forward Feature Selection:

In [135]:
from sklearn.feature_selection import f_regression
ffs = f_regression(X,Y)

variable = [ ]
for i in range(0,len(X.columns)-1):
    if ffs[0][i] >=10:
       variable.append(X.columns[i])

In [136]:
df_best = X[variable]

df_best.shape

(251, 391)

We have choosen f-value as 10. Based on the value, 391 features were selected out of 24711.

In [137]:
#converting all the datatypes of dataframe to float
df_best = df_best.astype(float)
Y = Y.astype(float)
df_best.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 251 entries, 0 to 250
Columns: 391 entries, 225209_s_at to 227068_at
dtypes: float64(391)
memory usage: 768.7 KB


## Principle component analysis & SVC:

Principle Component Analysis is used to reduce the dimension of the obtained feature(391).

PCA reduce the total features less than or equal to the samples.

In [138]:
from sklearn.ensemble import RandomForestRegressor

RFR_acc = []
val = []
acc =0
for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, Y, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  #Splitting the training data into training and validation data using K-Fold
  acc_score = []
  X_train1, X_test1, y_train1, y_test1 = train_test_split(X_train, y_train, train_size=0.7, random_state=11, shuffle=True) #Train Test Split
  
  forest_regressor = RandomForestRegressor(n_estimators = 300, random_state = 0)
  forest_regressor.fit(X_train1, y_train1)

  y_predict3 = forest_regressor.predict(X_test1)
  accuracy = r2_score(y_test1, y_predict3)
  acc_score.append(accuracy)
     
  RFR_acc.append(accuracy)
  val.append(i)
  if (acc < accuracy):
    acc = accuracy
    best_feat = i
    best_pca = x_pca

NameError: name 'r2_score' is not defined

## Visualising and selecting the best number of features

In [ ]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,RFR_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

### Predicting the Test Data

In [ ]:
r2_score(y_test,forest_regressor.predict(X_test))*100

Conclusion:

Model Performs in both Training and Testing data. But the accuracy is not impressive.

Required Number of features to achieve good accuracy = **30**

Training Accuracy =**24%**

Testing Accuracy = **6%**

In [ ]:
#XG Boost Regression model is used here for the classification
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from xgboost.sklearn import XGBRegressor
from sklearn.model_selection import KFold
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error

XGB_acc = []
val = []
acc =0
for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, Y, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  #Splitting the training data into training and validation data using K-Fold
  acc_score = []
  X_train1, X_test1, y_train1, y_test1 = train_test_split(X_train, y_train, train_size=0.7, random_state=11, shuffle=True) #Train Test Split
    
  XGBR = XGBRegressor(objective ='reg:squarederror')
  XGBR.fit(X_train1, y_train1)

  y_predict5 = XGBR.predict(X_test1)
  accuracy = r2_score(y_test1, y_predict5)
  acc_score.append(accuracy)
     
  XGB_acc.append(accuracy)
  val.append(i)
  if (acc < accuracy):
    acc = accuracy
    best_feat = i
    best_pca = x_pca

In [ ]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,XGB_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

In [ ]:
r2_score(y_test,XGBR.predict(X_test))*100

# 10. Target Variable - DSSTime:

In [ ]:
df = df_main.copy()
df1 =df_output.copy()

### Combining both Input data and Output data(DSSTime)

Merging all the input data from the dataframe "df" and "DSSTime" variable from the dataframe "df1"


In [ ]:
df_final = pd.merge(df,df1[['DSS.time']], left_on=['probe'],right_on=[df1['_PATIENT']],how='inner')

In [ ]:
df_final.head()

## Missing Values

Removing the missing values

In [ ]:
df_final.isna().sum()

Replaced missing values with the mean of the same target variable

In [ ]:
df_final['DSS.time'] = df_final['DSS.time'].replace(np.NaN,df_final['DSS.time'].mean())

In [ ]:
df_final.isna().sum()

In [ ]:
X = df_final.drop(columns = ['probe','DSS.time'],axis=1)
Y = df_final['DSS.time']


In [ ]:
print(X.shape, Y.shape)

We could see that the X shape is 251 rows and 24711 columns.
More columns can make the model to perform poor, so its better to reject the features, which is highly correlated. 

Here, we have choosen Forward Feature selection to select the best feature for the target variable - DSS.time.

### Feature Selection - Forward Feature Selection:

In [ ]:
from sklearn.feature_selection import f_regression
ffs = f_regression(X,Y)

variable = [ ]
for i in range(0,len(X.columns)-1):
    if ffs[0][i] >=10:
       variable.append(X.columns[i])

In [ ]:
df_best = X[variable]

df_best.shape

We have choosen f-value as 10. Based on the value, 963 features were selected out of 24711.

In [ ]:
#converting all the datatypes of dataframe to float
df_best = df_best.astype(float)
Y = Y.astype(float)
df_best.info()

## Principle component analysis & XG Boost Regressor:

Principle Component Analysis is used to reduce the dimension of the obtained feature(963).

PCA reduce the total features less than or equal to the samples.

In [ ]:
#XG Boost Regression model is used here for the Regression
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from xgboost.sklearn import XGBRegressor
from sklearn.model_selection import KFold
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error

XGB_acc = []
val = []
acc =0
for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, Y, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  acc_score = []
  X_train1, X_test1, y_train1, y_test1 = train_test_split(X_train, y_train, train_size=0.7, random_state=11, shuffle=True) #Train Test Split
    
  XGBR = XGBRegressor(objective ='reg:squarederror')
  XGBR.fit(X_train1, y_train1)

  y_predict5 = XGBR.predict(X_test1)
  accuracy = r2_score(y_test1, y_predict5)
  acc_score.append(accuracy)
     
  XGB_acc.append(accuracy)
  val.append(i)
  if (acc < accuracy):
    acc = accuracy
    best_feat = i
    best_pca = x_pca

## Visualising and selecting the best number of features

In [ ]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,XGB_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

### Predicting the Test Data

In [ ]:
r2_score(y_test,XGBR.predict(X_test))*100

Conclusion:

XGBoost Model Performs poor in both Training and Testing data. The accuracy is not impressive.

Required Number of features to achieve good accuracy = **30**

Training Accuracy =**16%**

Testing Accuracy = **-10%**

So, we can try with different model

In [ ]:
#Gradient Boosting Regression model is used here for the Regression
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from sklearn.svm import SVR

XGB_acc = []
val = []
acc =0
for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, Y, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  acc_score = []
  X_train1, X_test1, y_train1, y_test1 = train_test_split(X_train, y_train, train_size=0.7, random_state=11, shuffle=True) #Train Test Split
    

  svr_regressor = SVR(kernel='linear', C=10, gamma='auto')
  svr_regressor.fit(X_train1, y_train1)

  y_predict1 = svr_regressor.predict(X_test1)
  accuracy = r2_score(y_test1, y_predict1)
  acc_score.append(accuracy)
      
  XGB_acc.append(accuracy)
  val.append(i)
  if (acc < accuracy):
    acc = accuracy
    best_feat = i
    best_pca = x_pca

In [ ]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,XGB_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

In [ ]:
r2_score(y_test,svr_regressor.predict(X_test))*100

Conclusion:

SVR Model Performs both in Training and Testing data.

Required Number of features to achieve good accuracy = 240

Training Accuracy =32%

Testing Accuracy = 11%

# 11. Target Variable - AgeAtDiagnosis:

In [ ]:
df = df_main.copy()
df1 =df_output.copy()

### Combining both Input data and Output data(AgeAtDiagnosis)

Merging all the input data from the dataframe "df" and "AgeAtDiagnosis" variable from the dataframe "df1"


In [ ]:
df_final = pd.merge(df,df1[['AgeAtDiagnosis']], left_on=['probe'],right_on=[df1['_PATIENT']],how='inner')

In [ ]:
df_final.head()

## Missing Values

Checking if there is any missing values

In [ ]:
df_final.isna().sum()

In [ ]:
X = df_final.drop(columns = ['probe','AgeAtDiagnosis'],axis=1)
Y = df_final['AgeAtDiagnosis']


In [ ]:
print(X.shape, Y.shape)

We could see that the X shape is 251 rows and 24711 columns.
More columns can make the model to perform poor, so its better to reject the features, which is highly correlated. 

Here, we have choosen Forward Feature selection to select the best feature for the target variable - AgeAtDiagnosis.

### Feature Selection - Forward Feature Selection:

In [ ]:
from sklearn.feature_selection import f_regression
ffs = f_regression(X,Y)

variable = [ ]
for i in range(0,len(X.columns)-1):
    if ffs[0][i] >=10:
       variable.append(X.columns[i])

In [ ]:
df_best = X[variable]

df_best.shape

We have choosen f-value as 10. Based on the value, 1120 features were selected out of 24711.

In [ ]:
#converting all the datatypes of dataframe to float
df_best = df_best.astype(float)
Y = Y.astype(int)
df_best.info()

## Principle component analysis & Linear Regression:

Principle Component Analysis is used to reduce the dimension of the obtained feature(1120).

PCA reduce the total features less than or equal to the samples.

In [ ]:
from sklearn.linear_model import LinearRegression

RFR_acc = []
val = []
acc =0

for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, Y.values, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  acc_score = []
  kf = KFold(n_splits=k, shuffle=False, random_state=None)
  for train_index , test_index in kf.split(X_train):
    X_train1 , X_test1 = X_train[train_index,:],X_train[test_index,:]
    y_train1 , y_test1 = y_train[train_index] , y_train[test_index]
  
    linear_regressor = LinearRegression(normalize=True)
    linear_regressor.fit(X_train1, y_train1)

    y_predict = linear_regressor.predict(X_test1)
    accuracy = (r2_score(y_test1, y_predict))*100
    acc_score.append(accuracy)

  avg_acc_score = sum(acc_score)/k   
  RFR_acc.append(avg_acc_score)
  val.append(i)
  if (acc < avg_acc_score):
    acc = accuracy
    best_feat = i
    best_pca = x_pca

## Visualising and selecting the best number of features

In [ ]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,RFR_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

### Predicting the Test Data

In [ ]:
r2_score(y_test,linear_regressor.predict(X_test))*100

Conclusion:

Model Performs poor in both Training and Testing data. The accuracy is not impressive.

Required Number of features to achieve good accuracy = **240**

Training Accuracy =**2%**

Testing Accuracy = **-6%**

First Model seems to be overfitting, since the model performs good in Training but poor in Test. SO, we have implemented K-Fold cross validation, but the performance of model is not impressive.

Hence, lets try with different model to explore

In [ ]:
#Random Forest Regressor
from sklearn.ensemble import RandomForestRegressor

RFR_acc = []
val = []
acc =0
for i in range(10,250,10):

  pca = PCA(n_components = i)
  x_pca = pca.fit_transform(df_best)
  
  X_train, X_test, y_train, y_test = train_test_split(x_pca, Y, train_size=0.7, random_state=11, shuffle=True) #Train Test Split

  acc_score = []
  X_train1, X_test1, y_train1, y_test1 = train_test_split(X_train, y_train, train_size=0.7, random_state=11, shuffle=True) #Train Test Split
  
  forest_regressor = RandomForestRegressor(n_estimators = 300, random_state = 0)
  forest_regressor.fit(X_train1, y_train1)

  y_predict3 = forest_regressor.predict(X_test1)
  accuracy = r2_score(y_test1, y_predict3)
  acc_score.append(accuracy)
     
  RFR_acc.append(accuracy)
  val.append(i)
  if (acc < accuracy):
    acc = accuracy
    best_feat = i
    best_pca = x_pca

In [ ]:
plt.figure(figsize=(18,5))
#plt.bar(models,accuracies)
plt.xlabel("Number of Features")
plt.ylabel("Accuracy in %")
sns.barplot(val,RFR_acc)
print("Maximum Accuracy is ",acc, " Best number of features : ", best_feat)

In [ ]:
r2_score(y_test,forest_regressor.predict(X_test))*100

Conclusion:

Model Performs both in Training and Testing data. The Testing accuracy looks better than previous model.

Required Number of features to achieve good accuracy = **10**

Training Accuracy =**36%**

Testing Accuracy = **13%**